### ISSUE TITLE
Support deleting resources

### ISSUE THREAD

**Issue by sebastianroesch** [pusher/faros#92]
Currently, Faros supports creating and updating resources. Deleting resources requires manual work, for example by deleting the resource in Kubernetes as well as in the GitOps repository.

**Consideration:**
One possibility would be to compare the resources in the GitOps repository to the resources in the cluster managed by Faros and delete the ones that no longer exist in the GitOps repository. This could result in resources accidentally being deleted, be it by developers, a misconfiguration of Faros or other reasons.
A safer way would be to be more intentional about the resources that should be deleted, by marking them for deletion in the GitOps repository. 

**Proposal:**
Add an annotation to mark a resource to be deleted by Faros. 
Setting the `faros.pusher.com/resource-state` annotation to `marked-for-deletion` will remove the resource in the Kubernetes cluster on the next `Reconcile()`.
The annotation is optional and defaults to `active`, which would tell Faros to create or update the resource as it normally does.

Example:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: example
  namespace: default
  labels:
    app: nginx
  annotations:
    faros.pusher.com/resource-state: "marked-for-deletion"
```

**Comment by JoelSpeed** [comment]
> Currently, Faros supports creating and updating resources. Deleting resources requires manual work, for example by deleting the resource in Kubernetes as well as in the GitOps repository.

Faros is meant to track what is existing inside the repository and delete any resources existing on the cluster that have been deleted from the repository

https://github.com/pusher/faros/blob/master/pkg/controller/gittrack/gittrack_controller.go#L587

On each pass of the `GitTrack` controller it finds all `GitTrackObjects` owned by the `GitTrack` and compares this list with the files present in the repository. Therefore if you delete a resource definition from the repository (note this is keyed by name, namespace and kind so renaming a file is ok), the `GitTrack` controller deletes the `GitTrackObject` and in turn the Kubernetes Garbage Collector deletes the child resource.

> This could result in resources accidentally being deleted, be it by developers, a misconfiguration of Faros or other reasons.

We are aware there is a risk here, our current plan is to implement a Webhook Admission controller to prevent changes to the `repository` and `subpath` fields of the `GitTrack` once created, making it impossible for people to quickly modify the path within their repo to track (which could cause deletions) and force all changes within Git.

Are you not seeing this behaviour in your tests?

**Comment by sebastianroesch** [comment]
Thanks for the clarification. No, we did not see this behavior in our tests, but were able to reproduce it just now with a minimal setup. So we will look into why we were not able to see this in our evaluation. We're currently suspecting something in minikube or our test resources.

**Comment by sebastianroesch** [comment]
I'm closing this issue as resource deletion works as described here. We believe we did not notice resources being deleted as, in our tests, we only had one resource deployed. When deleted, the directory watched by Faros was empty, so Faros skipped applying changes.

As for the plans to reduce the risk of misconfiguration, I've opened up #101 where I summarised our concerns with deleting resources.
